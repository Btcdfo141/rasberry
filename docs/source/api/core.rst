.. _core_module:

:mod:`spencerassistant.core`
-------------------------

.. automodule:: spencerassistant.core
    :members: