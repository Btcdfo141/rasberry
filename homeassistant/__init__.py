"""Init file for Home Assistant."""


test2
