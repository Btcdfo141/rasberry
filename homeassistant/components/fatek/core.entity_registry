{
  "version": 1,
  "minor_version": 8,
  "key": "core.entity_registry",
  "data": {
    "entities": [
      {
        "area_id": null,
        "capabilities": null,
        "config_entry_id": null,
        "device_class": null,
        "device_id": null,
        "disabled_by": null,
        "entity_category": null,
        "entity_id": "person.marcin_ziarko",
        "hidden_by": null,
        "icon": null,
        "id": "0fa03fbe073a63325a3af0ae10563f01",
        "has_entity_name": false,
        "name": null,
        "options": {},
        "original_device_class": null,
        "original_icon": null,
        "original_name": "Marcin Ziarko",
        "platform": "person",
        "supported_features": 0,
        "unique_id": "marcin_ziarko",
        "unit_of_measurement": null
      },
      {
        "area_id": null,
        "capabilities": {
          "supported_color_modes": [
            "onoff"
          ]
        },
        "config_entry_id": null,
        "device_class": null,
        "device_id": null,
        "disabled_by": null,
        "entity_category": null,
        "entity_id": "light.m878",
        "hidden_by": null,
        "icon": null,
        "id": "379ff55d05771bf72f8ac457aa778ea3",
        "has_entity_name": false,
        "name": null,
        "options": {},
        "original_device_class": null,
        "original_icon": null,
        "original_name": "Poddasze TV",
        "platform": "fatek",
        "supported_features": 0,
        "unique_id": "M878",
        "unit_of_measurement": null
      }
    ]
  }
}
