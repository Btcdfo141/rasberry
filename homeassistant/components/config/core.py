"""Component to interact with Hassbian tools."""
import asyncio

from homeassistant.components.http import HomeAssistantView
from homeassistant.config import async_check_ha_config_file


@asyncio.coroutine
def async_setup(hass):
    """Set up the Hassbian config."""
    hass.http.register_view(CheckConfigView)
    return True


class CheckConfigView(HomeAssistantView):
    """Hassbian packages endpoint."""

    url = '/api/config/core/check_config'
    name = 'api:config:core:check_config'

    @asyncio.coroutine
    def post(self, request):
        """Validate configuration and return results."""
        hass = request.app['hass']

        if hass.components.hassio.is_hassio():
            errors = yield from hass.components.hassio.async_check_config()
        else:
            errors = yield from async_check_ha_config_file(hass)

        state = 'invalid' if errors else 'valid'

        return self.json({
            "result": state,
            "errors": errors,
        })
