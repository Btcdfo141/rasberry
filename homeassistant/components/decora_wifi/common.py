"""Common code for decora_wifi."""

from __future__ import annotations

from decora_wifi import DecoraWiFiSession
from decora_wifi.models.iot_switch import IotSwitch
from decora_wifi.models.person import Person
from decora_wifi.models.residence import Residence
from decora_wifi.models.residential_account import ResidentialAccount
from decora_wifi.models.residential_permission import ResidentialPermission

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers import device_registry
from homeassistant.helpers.entity import Entity

from .const import LIGHT_DOMAIN, PLATFORMS


class DecoraWifiError(HomeAssistantError):
    """Base for errors raised when the decora_wifi integration encounters an issue."""


class CommFailed(DecoraWifiError):
    """Raised when DecoraWifiPlatform.login() fails to communicate with the myLeviton Service."""


class SessionEntityNotFound(DecoraWifiError):
    """Raised if a platform fails to find the session entity during setup."""


class LoginFailed(DecoraWifiError):
    """Raised when DecoraWifiPlatform.login() fails to log in."""


class LoginMismatch(DecoraWifiError):
    """Raised if the userid returned on reauth does not match the userid cached when the integration was originally set up."""


class DecoraWifiPlatform:
    """Class to hold decora_wifi platform sessions and related methods."""

    def __init__(self, email: str, password: str) -> None:
        """Iniialize session holder."""
        self._session = None
        self._email = email
        self._name = f"Decora_Wifi - {self._email}"
        self._password = password
        self._iot_switches: dict[str, IotSwitch] = {}
        self._logged_in = False

    @property
    def active_platforms(self) -> list[str]:
        """Get the list of platforms which have devices defined."""
        return [p for p in PLATFORMS if self._iot_switches[p]]

    @property
    def lights(self) -> list[IotSwitch]:
        """Get the lights."""
        return self._iot_switches[LIGHT_DOMAIN]

    def _api_login(self):
        """Log in to decora_wifi session."""
        try:
            user = self._session.login(self._email, self._password)
            # If the call to the decora_wifi API's session.login returns None, there was a problem with the credentials.
            if user is None:
                raise LoginFailed
            self._logged_in = True
        except ValueError as exc:
            self._logged_in = False
            raise CommFailed from exc
        self._logged_in = True

    def _api_logout(self):
        """Log out of decora_wifi session."""
        if self._logged_in:
            try:
                Person.logout(self._session)
            except ValueError as exc:
                raise CommFailed from exc
        self._logged_in = False

    def _api_get_devices(self):
        """Update the device library from the API."""
        perms: list[ResidentialPermission] = []
        accounts: list[ResidentialAccount] = []
        residences: list[Residence] = []
        switches: list[IotSwitch] = []

        # Gather permissions
        try:
            perms.extend(self._session.user.get_residential_permissions())
        except ValueError as exc:
            self._logged_in = False
            raise CommFailed from exc

        # Gather residences for which the logged in user has permissions
        for permission in perms:
            if permission.residentialAccountId is not None:
                accounts.append(
                    ResidentialAccount(self._session, permission.residentialAccountId)
                )
            elif permission.residenceId is not None:
                residences.append(Residence(self._session, permission.residenceId))
        for acct in accounts:
            try:
                residences.extend(acct.get_residences())
            except ValueError as exc:
                raise CommFailed from exc

        # Gather switches from residences
        for res in residences:
            try:
                switches.extend(res.get_iot_switches())
            except ValueError as exc:
                raise CommFailed from exc

        # Add the switches to the appropriate list in the iot_switches dictionary.
        for switch in switches:
            platform = DecoraWifiPlatform.classify_device(switch)
            self._iot_switches[platform].append(switch)

    def refresh_devices(self):
        """Refresh this object's devices."""
        self._iot_switches: dict[str, IotSwitch] = {
            platform: [] for platform in PLATFORMS
        }
        self._api_get_devices()

    def setup(self):
        """Set up the session after object instantiation."""
        self._iot_switches = {platform: [] for platform in PLATFORMS}
        self._session = DecoraWiFiSession()
        self._api_login()
        self._api_get_devices()

    def teardown(self):
        """Clean up the session on object deletion."""
        self._api_logout()

    @staticmethod
    async def async_setup_decora_wifi(hass: HomeAssistant, email: str, password: str):
        """Set up a decora wifi session."""

        def setup_platform() -> DecoraWifiPlatform:
            platform = DecoraWifiPlatform(email, password)
            platform.setup()
            return platform

        return await hass.async_add_executor_job(setup_platform)

    @staticmethod
    def classify_device(dev):
        """Classify devices by platform."""
        # The light platform is the only one currently implemented in the integration.
        return LIGHT_DOMAIN


class DecoraWifiEntity(Entity):
    """Base Class for decora_wifi entities."""

    def __init__(self, device: IotSwitch) -> None:
        """Initialize Decora Wifi device base class."""
        self._switch = device
        self._model = device.model
        self._mac_address = device.mac

    @property
    def device_info(self):
        """Return device info for the associated device."""
        return {
            "name": self._switch.name,
            "connections": {
                (device_registry.CONNECTION_NETWORK_MAC, self._mac_address)
            },
            "manufacturer": self._switch.manufacturer,
            "model": self._model,
            "sw_version": self._switch.version,
        }

    @property
    def name(self):
        """Return the display name of this switch."""
        return self._switch.name

    @property
    def unique_id(self):
        """Return the unique id of the switch."""
        return self._mac_address
