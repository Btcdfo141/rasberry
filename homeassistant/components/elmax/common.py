"""Elmax integration common classes and utilities."""
from __future__ import annotations

from datetime import timedelta
import logging
from logging import Logger
import ssl

import async_timeout
from elmax_api.exceptions import (
    ElmaxApiError,
    ElmaxBadLoginError,
    ElmaxBadPinError,
    ElmaxNetworkError,
)
from elmax_api.http import GenericElmax
from elmax_api.model.actuator import Actuator
from elmax_api.model.area import Area
from elmax_api.model.endpoint import DeviceEndpoint
from elmax_api.model.panel import PanelEntry, PanelStatus
from httpx import ConnectError, ConnectTimeout
from packaging import version

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.update_coordinator import (
    CoordinatorEntity,
    DataUpdateCoordinator,
    UpdateFailed,
)

from .const import (
    DEFAULT_TIMEOUT,
    DOMAIN,
    ELMAX_LOCAL_API_PATH,
    MIN_APIV2_SUPPORTED_VERSION,
)

_LOGGER = logging.getLogger(__name__)


def get_direct_api_url(host: str, port: int, use_ssl: bool) -> str:
    """Return the direct API url given the base URI."""
    schema = "https" if use_ssl else "http"
    return f"{schema}://{host}:{port}/{ELMAX_LOCAL_API_PATH}"


def build_direct_ssl_context(cadata: str) -> ssl.SSLContext:
    """Create a custom SSL context for direct-api verification."""
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_verify_locations(cadata=cadata)
    return context


def check_local_version_supported(api_version: str | None) -> bool:
    """Check whether the given API version is supported."""
    if api_version is None:
        return False
    return version.parse(api_version) >= version.parse(MIN_APIV2_SUPPORTED_VERSION)


class DummyPanel(PanelEntry):
    """Helper class for wrapping a directly accessed Elmax Panel."""

    def __init__(self, panel_uri):
        """Construct the object."""
        super().__init__(panel_uri, True, {})

    def get_name_by_user(self, username: str) -> str:
        """Return the panel name."""
        return f"Direct Panel {self.hash}"


class ElmaxCoordinator(DataUpdateCoordinator[PanelStatus]):
    """Coordinator helper to handle Elmax API polling."""

    def __init__(
        self,
        hass: HomeAssistant,
        logger: Logger,
        elmax_api_client: GenericElmax,
        panel: PanelEntry,
        name: str,
        update_interval: timedelta,
    ) -> None:
        """Instantiate the object."""
        self._client = elmax_api_client
        self._panel_entry = panel
        self._state_by_endpoint = None
        super().__init__(
            hass=hass, logger=logger, name=name, update_interval=update_interval
        )

    @property
    def panel_entry(self) -> PanelEntry:
        """Return the panel entry."""
        return self._panel_entry

    def get_actuator_state(self, actuator_id: str) -> Actuator:
        """Return state of a specific actuator."""
        if self._state_by_endpoint is not None:
            return self._state_by_endpoint[actuator_id]
        raise HomeAssistantError("Unknown actuator")

    def get_zone_state(self, zone_id: str) -> Actuator:
        """Return state of a specific zone."""
        if self._state_by_endpoint is not None:
            return self._state_by_endpoint[zone_id]
        raise HomeAssistantError("Unknown zone")

    def get_area_state(self, area_id: str) -> Area:
        """Return state of a specific area."""
        if self._state_by_endpoint is not None and area_id:
            return self._state_by_endpoint[area_id]
        raise HomeAssistantError("Unknown area")

    @property
    def http_client(self):
        """Return the current http client being used by this instance."""
        return self._client

    @http_client.setter
    def http_client(self, client: GenericElmax):
        """Set the client library instance for Elmax API."""
        self._client = client

    async def _async_update_data(self):
        try:
            async with async_timeout.timeout(DEFAULT_TIMEOUT):
                # The following command might fail in case of the panel is offline.
                # In this case, just print a warning and return None: listeners will assume the panel
                # offline.
                status = await self._client.get_current_panel_status()

                # Store a dictionary for fast endpoint state access
                self._state_by_endpoint = {
                    k.endpoint_id: k for k in status.all_endpoints
                }
                return status

        except ElmaxBadPinError as err:
            raise ConfigEntryAuthFailed("Control panel pin was refused") from err
        except ElmaxBadLoginError as err:
            raise ConfigEntryAuthFailed("Refused username/password/pin") from err
        except ElmaxApiError as err:
            raise UpdateFailed(f"Error communicating with ELMAX API: {err}") from err
        except (ConnectError, ConnectTimeout, ElmaxNetworkError) as err:
            raise UpdateFailed(
                "A network error occurred while communicating with Cloud/Elmax Panel."
                "If connecting against the Cloud, make sure HA can reach the internet."
                "If connecting directly to the Elmax Panel, make sure the panel is online and "
                "no firewall is blocking it."
            ) from err


class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
    """Wrapper for Elmax entities."""

    def __init__(
        self,
        elmax_device: DeviceEndpoint,
        panel_version: str,
        coordinator: ElmaxCoordinator,
    ) -> None:
        """Construct the object."""
        super().__init__(coordinator=coordinator)
        self._device = elmax_device
        self._panel_version = panel_version
        self._client = coordinator.http_client

    @property
    def panel_id(self) -> str | None:
        """Retrieve the panel id."""
        return self.coordinator.panel_entry.hash

    @property
    def unique_id(self) -> str | None:
        """Provide a unique id for this entity."""
        return self._device.endpoint_id

    @property
    def name(self) -> str | None:
        """Return the entity name."""
        return self._device.name

    @property
    def device_info(self) -> DeviceInfo | None:
        """Return device specific attributes."""
        return {
            "identifiers": {(DOMAIN, self.coordinator.panel_entry.hash)},
            "name": self.coordinator.panel_entry.get_name_by_user(
                self.coordinator.http_client.get_authenticated_username()
            ),
            "manufacturer": "Elmax",
            "model": self._panel_version,
            "sw_version": self._panel_version,
        }

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.coordinator.panel_entry.online
