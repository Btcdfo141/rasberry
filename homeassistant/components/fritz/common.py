"""Support for AVM Fritz!Box classes."""
from collections import namedtuple
from datetime import datetime, timedelta
import logging
from typing import Any, Dict, Optional

# pylint: disable=import-error
from fritzconnection import FritzConnection
from fritzconnection.core.exceptions import FritzConnectionException
from fritzconnection.lib.fritzhosts import FritzHosts
from fritzconnection.lib.fritzstatus import FritzStatus
import voluptuous as vol

from homeassistant.const import (
    CONF_DEVICES,
    CONF_HOST,
    CONF_PASSWORD,
    CONF_PORT,
    CONF_USERNAME,
)
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.dispatcher import async_dispatcher_send
from homeassistant.helpers.event import async_track_time_interval
from homeassistant.util import dt as dt_util, get_local_ip

from .const import (
    ATTR_HOST,
    DEFAULT_HOST,
    DEFAULT_PORT,
    DEFAULT_USERNAME,
    DOMAIN,
    ERROR_CONNECTION_ERROR,
    TRACKER_SCAN_INTERVAL,
)

_LOGGER = logging.getLogger(__name__)

Device = namedtuple("Device", ["mac", "ip", "name"])


CONFIG_SCHEMA = vol.Schema(
    vol.All(
        cv.deprecated(DOMAIN),
        {
            DOMAIN: vol.Schema(
                {
                    vol.Required(CONF_DEVICES): vol.All(
                        cv.ensure_list,
                        [
                            vol.Schema(
                                {
                                    vol.Optional(CONF_HOST): cv.string,
                                    vol.Optional(CONF_PORT): cv.port,
                                    vol.Required(CONF_USERNAME): cv.string,
                                    vol.Required(CONF_PASSWORD): cv.string,
                                }
                            )
                        ],
                    )
                }
            )
        },
    ),
    extra=vol.ALLOW_EXTRA,
)

SERVICE_SCHEMA = vol.Schema({vol.Required(ATTR_HOST): cv.string})


class FritzBoxTools:
    """FrtizBoxTools class."""

    def __init__(
        self,
        hass,
        password,
        username=DEFAULT_USERNAME,
        host=DEFAULT_HOST,
        port=DEFAULT_PORT,
    ):
        """Initialize FritzboxTools class."""
        self._cancel_scan = None
        self._device_info = None
        self._devices: Dict[str, Any] = {}
        self._unique_id = None
        self.connection = None
        self.error = None
        self.fritzhosts = None
        self.fritzstatus = None
        self.ha_ip = None
        self.hass = hass
        self.host = host
        self.password = password
        self.port = port
        self.success = None
        self.username = username

    async def async_setup(self):
        """Wrap up FritzboxTools class setup."""
        return await self.hass.async_add_executor_job(self.setup)

    def setup(self):
        """Set up FritzboxTools class."""
        try:
            self.connection = FritzConnection(
                address=self.host,
                port=self.port,
                user=self.username,
                password=self.password,
                timeout=60.0,
            )

            self.fritzstatus = FritzStatus(fc=self.connection)
            self._unique_id = self.connection.call_action("DeviceInfo:1", "GetInfo")[
                "NewSerialNumber"
            ]
            self.fritzhosts = FritzHosts(fc=self.connection)
            self._device_info = self._fetch_device_info()
            self.success = True
            self.error = False
        except FritzConnectionException:
            self.success = False
            self.error = ERROR_CONNECTION_ERROR

        self.ha_ip = get_local_ip()

        self.scan_devices()

        self._cancel_scan = async_track_time_interval(
            self.hass, self.scan_devices, timedelta(seconds=TRACKER_SCAN_INTERVAL)
        )

        return self.success, self.error

    def unload(self):
        """Unload FritzboxTools class."""
        _LOGGER.debug("Unloading Fritz!Box router integration")
        if self._cancel_scan is not None:
            self._cancel_scan()
            self._cancel_scan = None

    @property
    def unique_id(self):
        """Return unique id."""
        return self._unique_id

    @property
    def fritzbox_model(self):
        """Return model."""
        return self._device_info["model"].replace("FRITZ!Box ", "")

    @property
    def device_info(self):
        """Return device info."""
        return self._device_info

    @property
    def devices(self) -> Dict[str, Any]:
        """Return devices."""
        return self._devices

    @property
    def signal_device_new(self) -> str:
        """Event specific per Fritzbox entry to signal new device."""
        return f"{DOMAIN}-device-new-{self._unique_id}"

    @property
    def signal_device_update(self) -> str:
        """Event specific per Fritzbox entry to signal updates in devices."""
        return f"{DOMAIN}-device-update-{self._unique_id}"

    def _update_info(self):
        """Retrieve latest information from the FRITZ!Box."""
        if not self.success:
            return None

        return self.fritzhosts.get_hosts_info()

    def scan_devices(self, now: Optional[datetime] = None) -> None:
        """Scan for new devices and return a list of found device ids."""

        _LOGGER.debug("Checking devices for Fritz!Box router %s", self.host)

        new_device = False
        for known_host in self._update_info():
            if not known_host.get("mac"):
                continue

            dev_mac = known_host["mac"]
            dev_name = known_host["name"]
            dev_ip = known_host["ip"]
            dev_home = known_host["status"]

            dev_info = Device(dev_mac, dev_ip, dev_name)

            if dev_mac in self._devices:
                self._devices[dev_mac].update(dev_info, dev_home)
            else:
                device = FritzDevice(dev_mac)
                device.update(dev_info, dev_home)
                self._devices[dev_mac] = device
                new_device = True

        async_dispatcher_send(self.hass, self.signal_device_update)
        if new_device:
            async_dispatcher_send(self.hass, self.signal_device_new)

    def _fetch_device_info(self):
        """Fetch device info."""
        info = self.connection.call_action("DeviceInfo:1", "GetInfo")
        return {
            "identifiers": {
                # Serial numbers are unique identifiers within a specific domain
                (DOMAIN, self.unique_id)
            },
            "name": info.get("NewModelName"),
            "manufacturer": "AVM",
            "model": info.get("NewModelName"),
            "sw_version": info.get("NewSoftwareVersion"),
        }


class FritzDevice:
    """FritzScanner device."""

    def __init__(self, mac, name=None):
        """Initialize device info."""
        self._mac = mac
        self._name = name
        self._ip_address = None
        self._last_activity = None
        self._connected = False

    def update(self, dev_info, dev_home):
        """Update device info."""
        utc_point_in_time = dt_util.utcnow()
        if not self._name:
            self._name = dev_info.name or self._mac.replace(":", "_")
        self._connected = dev_home

        if not self._connected:
            self._ip_address = None
        else:
            self._last_activity = utc_point_in_time
            self._ip_address = dev_info.ip

    @property
    def is_connected(self):
        """Return connected status."""
        return self._connected

    @property
    def mac_address(self):
        """Get MAC address."""
        return self._mac

    @property
    def hostname(self):
        """Get Name."""
        return self._name

    @property
    def ip_address(self):
        """Get IP address."""
        return self._ip_address

    @property
    def last_activity(self):
        """Return device last activity."""
        return self._last_activity
