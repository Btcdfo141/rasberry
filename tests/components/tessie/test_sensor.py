"""Test the Tessie sensor platform."""
from syrupy import SnapshotAssertion

from homeassistant.core import HomeAssistant

from .common import setup_platform


async def test_sensors(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Tests that the sensor entities are correct."""

    assert len(hass.states.async_all("sensor")) == 0

    await setup_platform(hass)

    assert hass.states.async_all("sensor") == snapshot(name="all")
