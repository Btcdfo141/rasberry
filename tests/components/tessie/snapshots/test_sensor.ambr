# serializer version: 1
# name: test_sensors[all]
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'battery',
        'friendly_name': 'Test Battery level',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': '%',
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_battery_level',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '75',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'energy',
        'friendly_name': 'Test Charge energy added',
        'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
        'unit_of_measurement': <UnitOfEnergy.KILO_WATT_HOUR: 'kWh'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_charge_energy_added',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '18.47',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'power',
        'friendly_name': 'Test Charger power',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPower.KILO_WATT: 'kW'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_charger_power',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '7',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'voltage',
        'friendly_name': 'Test Charger voltage',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfElectricPotential.VOLT: 'V'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_charger_voltage',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '224',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'current',
        'friendly_name': 'Test Charger current',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfElectricCurrent.AMPERE: 'A'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_charger_current',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '32',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'speed',
        'friendly_name': 'Test Charge rate',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfSpeed.KILOMETERS_PER_HOUR: 'km/h'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_charge_rate',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '49.2',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'distance',
        'friendly_name': 'Test Battery range',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_battery_range',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '424.35182592',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'speed',
        'friendly_name': 'Test Speed',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfSpeed.KILOMETERS_PER_HOUR: 'km/h'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_speed',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'unknown',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'power',
        'friendly_name': 'Test Power',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPower.KILO_WATT: 'kW'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_power',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '-7',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'enum',
        'friendly_name': 'Test Shift state',
        'icon': 'mdi:car-shift-pattern',
        'options': list([
          'p',
          'd',
          'r',
          'n',
        ]),
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_shift_state',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'unknown',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'distance',
        'friendly_name': 'Test Odometer',
        'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
        'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_odometer',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '8778.15941765875',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'pressure',
        'friendly_name': 'Test Tire pressure front left',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPressure.PSI: 'psi'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_tire_pressure_front_left',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '43.1487288094417',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'pressure',
        'friendly_name': 'Test Tire pressure front right',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPressure.PSI: 'psi'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_tire_pressure_front_right',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '43.1487288094417',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'pressure',
        'friendly_name': 'Test Tire pressure rear left',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPressure.PSI: 'psi'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_tire_pressure_rear_left',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '42.7861344496985',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'pressure',
        'friendly_name': 'Test Tire pressure rear right',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfPressure.PSI: 'psi'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_tire_pressure_rear_right',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '42.7861344496985',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'temperature',
        'friendly_name': 'Test Inside temperature',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfTemperature.CELSIUS: '째C'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_inside_temperature',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '30.4',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'temperature',
        'friendly_name': 'Test Outside temperature',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfTemperature.CELSIUS: '째C'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_outside_temperature',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '30.5',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'temperature',
        'friendly_name': 'Test Driver temperature setting',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfTemperature.CELSIUS: '째C'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_driver_temperature_setting',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '22.5',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'device_class': 'temperature',
        'friendly_name': 'Test Passenger temperature setting',
        'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
        'unit_of_measurement': <UnitOfTemperature.CELSIUS: '째C'>,
      }),
      'context': <ANY>,
      'entity_id': 'sensor.test_passenger_temperature_setting',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': '22.5',
    }),
  ])
# ---
